lib/reporting/protocols_report.rb

```diff
@@ -0,0 +1,196 @@
+# frozen_string_literal: true
+
+require 'csv'
+begin
+  require 'reporting/cloudwatch_client'
+  require 'reporting/cloudwatch_query_quoting'
+  require 'reporting/command_line_options'
+rescue LoadError => e
+  warn 'could not load paths, try running with "bundle exec rails runner"'
+  raise e
+end
+
+module Reporting
+  class ProtocolsReport
+    include Reporting::CloudwatchQueryQuoting
+
+    attr_reader :time_range
+
+    SAML_AUTH_EVENT = 'SAML Auth'
+    OIDC_AUTH_EVENT = 'OpenID Connect: authorization request'
+
+    # @param [Range<Time>] time_range
+    def initialize(
+      issuers:,
+      time_range:,
+      verbose: false,
+      progress: false,
+      slice: 1.day,
+      threads: 10
+    )
+      @issuers = issuers
+      @time_range = time_range
+      @verbose = verbose
+      @progress = progress
+      @slice = slice
+      @threads = threads
+    end
+
+    def verbose?
+      @verbose
+    end
+
+    def progress?
+      @progress
+    end
+
+    def as_tables
+      [
+        overview_table,
+        protocols_table,
+        saml_signature_issues_table,
+      ]
+    end
+
+    def as_emailable_reports
+      [
+        Reporting::EmailableReport.new(
+          title: 'Overview',
+          table: overview_table,
+        ),
+        Reporting::EmailableReport.new(
+          title: 'State of Authentication',
+          table: protocols_table,
+        ),
+        Reporting::EmailableReport.new(
+          title: 'SAML Signature Issues',
+          table: saml_signature_issues_table,
+        ),
+      ]
+    end
+
+    def to_csvs
+      as_tables.map do |table|
+        CSV.generate do |csv|
+          table.each do |row|
+            csv << row
+          end
+        end
+      end
+    end
+
+    # @return Array<Hash>
+    def data
+      @data ||= begin
+        fetch_results
+      end
+    end
+
+    def fetch_results
+      cloudwatch_client.fetch(query:, from: time_range.begin, to: time_range.end)
+    end
+
+    def query
+      params = {
+        event: quote([SAML_AUTH_EVENT, OIDC_AUTH_EVENT]),
+      }
+
+      format(<<~QUERY, params)
+        fields
+          name as protocol,
+          properties.event_properties.service_provider as issuer,
+          properties.event_properties.request_signed = 1 AS signed,
+          properties.event_properties.request_signed != 1 AS not_signed,
+          isempty(properties.event_properties.matching_cert_serial) and signed as invalid_signature
+        | filter name IN %{event}
+        | stats
+          count(*) as request_count,
```

Comment:
as mentioned in some of our standups, i think this report will end up growing beyond these initial needs, that will require different fields/manipulation of the events. so i was originally envisioning this to be a query for the events with some specific fields that are then manipulated in the `data` method (example: https://github.com/18F/identity-idp/blob/main/lib/reporting/drop_off_report.rb#L410-L421). so this will probably change in the future, but i think it's fine for now.

Reply:
Another option is also to have multiple separate queries that get combined in one email or something

Reply:
I have done that by pulling out the LOA ACR metrics that was also asked for in this ticket into a separate ticket (and consequently separate report). I do see requests by protocols and SAML signature issues as being separate concerns. I didn't want to do too much given the Data Warehouse plan. I also see that some of these reports are perfectly fine as CloudWatch reports run from the AWS web console...

Reply:
Decided to go with the separate queries after all (rationale - would have increased code duplication with separate reports). Will also be bringing in the LOA ACR query into this report.